Link developers and organizations to instances

rocioar opened the pull request:
Main changes:

- Adds Organization model
- Adds github_handle and organization to UserProfile
- Adds organization owner and user creator to the instanceReference
- Updated watch PRs task to watch PRs for all users registered in OCIM on the Watched forks.

Reviewers:

@itsjeyd

itsjeyd commented on userprofile/models.py:
```diff
@@ -31,12 +31,34 @@
+    Organizations model
+    """
+    name = models.CharField(max_length=255)
+    github_handle = models.CharField(unique=True, max_length=255)
```
@rocioar This field isn't used anywhere at the moment, so I think it would be fine to make it optional. Right now I'm thinking that we'll mostly be using it for informational purposes (and won't strictly need it to be able to implement the first batch of access control and billing features). But that might change in the future.

On a related note, it would be good to establish a direct connection between watched forks and organizations (by changing `WatchedFork.organization` to point to the `Organization` model), to prepare for implementing the billing features. Right now, the only way for us to relate an organization to a watched fork would be via the `WatchedFork.organization` and `Organization.github_handle` fields. But that will break down if an organization (a/k/a a client) would like to use PR watch functionality for multiple GitHub organizations.

rocioar replied:
I made the field optional.

Make sense to update the WatchedFork to point to the new organization model, I will update it.

rocioar commented on pr_watch/tasks.py:
```diff
@@ -50,8 +52,15 @@ def watch_pr():
+                    'github_username',
+                    flat=True
+                )
+            )
```
@itsjeyd one thing I'm wondering here, if we filter by only the organization that owns the fork, then wouldn't we be limiting the capabilities? Let's say we are watching "edx-platform" (the main repository), then one of us makes a PR, we are not the owners of the fork, so the OCIM wouldn't see it and there will not be any sandbox created.

itsjeyd replied:
@rocioar Good question! If you look at the [list of watched forks](https://console.opencraft.com/admin/pr_watch/watchedfork/), you'll see that there are multiple ones that target `edx/edx-platform`. Each one of those maps `edx/edx-platform` to a different organization, but none of these organizations actually owns `edx/edx-platform`. Conceptually it doesn't matter who owns a given fork -- OCIM just needs to know which PRs against that fork it should create sandboxes for. And we store that information in the `organization` field of the `WatchedFork` model. In other words, the way to "read" a `WatchedFork` is: OCIM will create sandboxes for PRs against `WatchedFork.fork` for all members of `WatchedFork.organization` (*not* "`WatchedFork.organization` owns `WatchedFork.fork`") 🙂

rocioar commented on userprofile/models.py:
```diff
@@ -31,12 +31,34 @@
+    pr_watch_active = models.BooleanField(
+        default=True,
+        help_text="Is PR watch active for this organization?"
+    )
```
Makes sense, thanks for the detailed explanation.

rocioar commented on pr_watch/models.py:
```diff
@@ -297,6 +300,10 @@ def update_instance_from_pr(self, pr):
             self.watched_fork.configuration_extra_settings,
             pr.extra_settings
         )
+        if not instance.ref.creator or instance.ref.owner:
```
Good catch! fixed

rocioar commented on instance/migrations/0106_auto_20180821_1605.py:
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
```
👍 done

rocioar commented on userprofile/migrations/0003_auto_20180821_1605.py:
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
```
👍 done

itsjeyd commented on pr_watch/models.py:
```diff
@@ -56,11 +56,14 @@ class WatchedFork(models.Model):
-        help_text=("Watched GitHub organization. E.g.: open-craft. PRs against the watched fork "
-                   "made by members of this organization will trigger a sandbox build"),
+        help_text=(
+            "Watched GitHub organization. E.g.: open-craft. PRs against the watched fork "
```
@rocioar Could you update this to reflect the conceptual changes that we are making to the logic for watching PRs? You could simplify the first part to something like:

> Organization to watch. PRs against against the watched fork ..."

rocioar replied:
👍  done

itsjeyd commented on pr_watch/migrations/0009_organization_db_index.py:
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
```
@rocioar Would it be possible to merge this migration into the previous one? (`update_organization`)?

rocioar replied:
@itsjeyd actually this was done because there was an issue with postgres when changing the type of a db_index from a char to an int, see: https://code.djangoproject.com/ticket/27860

itsjeyd replied:
@rocioar Ah, OK. Totally fine to keep both of these migrations, then :)

rocioar commented:
@itsjeyd I've addressed your feedback, and now fixed the issue with the buckets (reason why the build was failing, we were on the max amount of buckets that we could have).

itsjeyd commented on pr_watch/models.py:
```diff
@@ -61,7 +61,7 @@ class WatchedFork(models.Model):
         db_index=True,
         help_text=(
-            "Watched GitHub organization. E.g.: open-craft. PRs against the watched fork "
+            "Organization to watch. PRs against against the watched fork."
```
@rocioar Ack... You'll need to remove the full stop at the end of this line. Sorry if my [earlier comment](https://github.com/open-craft/opencraft/pull/346#discussion_r212939789) was confusing in that regard?

Also, changing the help text should generate a new migration?

rocioar replied:
@itsjeyd my mistake here, I did it in a rush. Fixed it now, and updated the migrations too (squashed with the other migration)

itsjeyd commented on pr_watch/models.py:
```diff
@@ -53,11 +56,13 @@ class WatchedFork(models.Model):
-        help_text=("Watched GitHub organization. E.g.: open-craft. PRs against the watched fork "
-                   "made by members of this organization will trigger a sandbox build"),
+        help_text=(
+            "Organization to watch. PRs against against the watched fork"
```
@rocioar But now this is missing the second part of the sentence... To be completely clear, here's what I think this help text should say:

> Organization to watch. PRs against the watched fork made by members of this organization will trigger a sandbox build.

rocioar commented:
@itsjeyd fixed tests. Will deploy to stage once tests finish running.

rocioar commented on pr_watch/migrations/0008_update_organization.py:
```diff
@@ -0,0 +1,26 @@
+        migrations.AddField(
+            model_name='watchedfork',
+            name='organization',
+            field=models.ForeignKey(null=True, help_text='Organization to watch. PRs against the watched fork made by members of this organization will trigger a sandbox build.', on_delete=django.db.models.deletion.CASCADE, to='userprofile.Organization'),
```
This was failing on stage, because there was data for the index. So I had to alter the migrations, so that it first removes the organization field completely, and then adds it again.

itsjeyd replied:
@rocioar Ah, OK -- so this migration (`pr_watch/migrations/0008_update_organization.py`) now takes care of removing the `organization` field from the `WatchedFork` model and adding it back`. What's the purpose of `pr_watch/migrations/0009_organization_db_index.py`, do we still need it?
